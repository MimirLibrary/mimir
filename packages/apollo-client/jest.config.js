module.exports = {
  displayName: 'apollo-client',
  preset: '../../jest.preset.js',
  coverageDirectory: '../../coverage/packages/apollo-client',
};
