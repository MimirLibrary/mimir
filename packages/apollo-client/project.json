{
  "sourceRoot": "packages/apollo-client/src",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/apollo-client",
        "main": "packages/apollo-client/src/index.ts",
        "tsConfig": "packages/apollo-client/tsconfig.lib.json",
        "assets": ["packages/apollo-client/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/apollo-client/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/apollo-client"],
      "options": {
        "jestConfig": "packages/apollo-client/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
