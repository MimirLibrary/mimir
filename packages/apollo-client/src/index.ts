export * from './lib/instance';
export * from './lib/generated/client-types';
