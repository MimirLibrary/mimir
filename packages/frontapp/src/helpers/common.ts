export const shortenText = (text: string, numberToShorten: number) => {
  return text.trim().slice(0, numberToShorten);
};
