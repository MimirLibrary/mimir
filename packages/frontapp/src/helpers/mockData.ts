export const mockStatus = [
  {
    person_id: 1,
    status: 'Free',
    created_at: '2022-09-08T03:58:39.228Z',
    id: 1,
    material_id: 1,
  },
];

export const mockMaterial = [
  {
    __typename: 'Material',
    author: 'test Author',
    category: 'test Category',
    created_at: '2022-09-08T03:58:39.228Z',
    description: 'Test Description',
    id: '1',
    id_type: 'Test Id_Type',
    identifier: '1111111111',
    is_donated: false,
    notification: [],
    picture: '',
    statuses: mockStatus,
    title: 'test',
    type: 'Book',
    updated_at: '2022-09-08T03:58:39.228Z',
  },
];
