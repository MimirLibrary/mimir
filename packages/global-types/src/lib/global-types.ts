
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */
export enum Permissions {
    GRANT_REVOKE_MANAGER = "GRANT_REVOKE_MANAGER"
}

export interface CreateStateInput {
    person_id: number;
    state: boolean;
    description?: Nullable<string>;
}

export interface BookInput {
    identifier: string;
    person_id: number;
}

export interface ProlongTimeInput {
    person_id: number;
    material_id: number;
}

export interface CreateLocationInput {
    location: string;
}

export interface RemoveLocationInput {
    location_id: number;
}

export interface CreateMaterialInput {
    identifier: string;
    id_type: string;
    location_id: number;
    type: string;
    title: string;
    picture?: Nullable<string>;
    author: string;
    category: string;
    description: string;
}

export interface RemoveMaterialInput {
    identifier: string;
    location_id: number;
    type: string;
}

export interface UpdateMaterialInput {
    identifier?: Nullable<string>;
    id_type?: Nullable<string>;
    location_id?: Nullable<number>;
    type?: Nullable<string>;
    title?: Nullable<string>;
    picture?: Nullable<string>;
    author?: Nullable<string>;
    category?: Nullable<string>;
    updated_at: DateTime;
    description?: Nullable<string>;
}

export interface SearchOneMaterial {
    identifier: string;
    locations?: Nullable<number[]>;
}

export interface DonateBookInput {
    identifier: string;
    id_type: string;
    location_id: number;
    type: string;
    title: string;
    author: string;
    category: string;
    description?: Nullable<string>;
    picture?: Nullable<string>;
    person_id: number;
    role: string;
    is_donated: boolean;
}

export interface SearchInput {
    search?: Nullable<string>;
    locations?: Nullable<number[]>;
}

export interface CreateMessageInput {
    title: string;
    message: string;
    material_id?: Nullable<number>;
    person_id: number;
    location_id: number;
}

export interface CreateAnswerNotification {
    id: number;
    person_id: number;
    message: string;
}

export interface CreateNotificationInput {
    material_id: number;
    person_id: number;
}

export interface CreateSimpleNotification {
    person_id: number;
    message: string;
}

export interface RemoveNotificationInput {
    material_id: number;
    person_id: number;
}

export interface CreatePersonInput {
    smg_id: string;
    location_id: number;
    type: string;
}

export interface UpdatePersonLocationInput {
    person_id: number;
    location_id: number;
}

export interface CreateStatusInput {
    material_id: number;
    person_id: number;
    status: string;
}

export interface BlockedUsers {
    id: string;
    description?: Nullable<string>;
    state: boolean;
    created_at: DateTime;
    person: Person;
}

export interface IQuery {
    getBlocksByPerson(person_id: string): Nullable<Nullable<BlockedUsers>[]> | Promise<Nullable<Nullable<BlockedUsers>[]>>;
    getReasonOfBlock(person_id: string): Nullable<BlockedUsers> | Promise<Nullable<BlockedUsers>>;
    getAllTakenItems(person_id: number): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getItemsForClaimHistory(person_id: number): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getAllLocations(): Nullable<Location>[] | Promise<Nullable<Location>[]>;
    getAllMaterials(locations?: Nullable<number[]>, limit?: Nullable<number>, offset?: Nullable<number>): Nullable<Material>[] | Promise<Nullable<Material>[]>;
    getMaterialById(id: string): Material | Promise<Material>;
    searchOfMaterials(input: SearchInput): Nullable<Nullable<Material>[]> | Promise<Nullable<Nullable<Material>[]>>;
    getMaterialByIdentifier(input: SearchOneMaterial): Material | Promise<Material>;
    getMaterialByIdentifierFromMetadata(identifier: string): Nullable<IMetaOfMaterial> | Promise<Nullable<IMetaOfMaterial>>;
    getAllDonatedMaterialsByPerson(id: string): Nullable<Nullable<Material>[]> | Promise<Nullable<Nullable<Material>[]>>;
    getMessagesByPerson(person_id: string): Nullable<Nullable<Message>[]> | Promise<Nullable<Nullable<Message>[]>>;
    getAllMessages(location_id: number): Nullable<Message[]> | Promise<Nullable<Message[]>>;
    getNotificationsByPerson(person_id: number): Nullable<Notification>[] | Promise<Nullable<Notification>[]>;
    getNotificationsByMaterial(material_id: number): Nullable<Notification>[] | Promise<Nullable<Notification>[]>;
    getOnePerson(id: string): Person | Promise<Person>;
    getAllPersons(username?: Nullable<string>, locations?: Nullable<number[]>): Person[] | Promise<Person[]>;
    getStatusesByPerson(person_id: string): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getStatusesByMaterial(material_id: string): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getAllStatusesIsOverdue(locations?: Nullable<number[]>): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    welcome(): string | Promise<string>;
}

export interface IMutation {
    createState(input: CreateStateInput): Nullable<BlockedUsers> | Promise<Nullable<BlockedUsers>>;
    claimBook(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    returnItem(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    rejectItem(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    prolongClaimPeriod(input?: Nullable<ProlongTimeInput>): BookUnionResult | Promise<BookUnionResult>;
    createLocation(input: CreateLocationInput): Nullable<Location> | Promise<Nullable<Location>>;
    removeLocation(input: RemoveLocationInput): Nullable<Location> | Promise<Nullable<Location>>;
    createMaterial(input: CreateMaterialInput): Material | Promise<Material>;
    removeMaterial(input: RemoveMaterialInput): Material | Promise<Material>;
    updateMaterial(input: UpdateMaterialInput): Material | Promise<Material>;
    donateBook(input: DonateBookInput): Material | Promise<Material>;
    createMessageForManager(input: CreateMessageInput): Message | Promise<Message>;
    createNotification(input: CreateNotificationInput): Nullable<Notification> | Promise<Nullable<Notification>>;
    removeNotification(input: RemoveNotificationInput): Nullable<Notification> | Promise<Nullable<Notification>>;
    createAnswerNotification(input: CreateAnswerNotification): Notification | Promise<Notification>;
    createSimpleNotification(input?: Nullable<CreateSimpleNotification>): Notification | Promise<Notification>;
    createPerson(input: CreatePersonInput): Person | Promise<Person>;
    addPersonLocation(input: UpdatePersonLocationInput): Person | Promise<Person>;
    removePersonLocation(input: UpdatePersonLocationInput): Person | Promise<Person>;
    changePersonRole(person_id: number, type: string): Person | Promise<Person>;
    createStatus(input: CreateStatusInput): Status | Promise<Status>;
}

export interface Location {
    id: string;
    location: string;
    persons?: Nullable<Nullable<Person>[]>;
    materials?: Nullable<Nullable<Material>[]>;
}

export interface Material {
    id: string;
    identifier: string;
    id_type: string;
    type: string;
    location_id: number;
    created_at: DateTime;
    updated_at: DateTime;
    title: string;
    picture?: Nullable<string>;
    author: string;
    category: string;
    location: Location;
    statuses: Nullable<Status>[];
    notifications: Nullable<Notification>[];
    messages: Nullable<Message>[];
    description: string;
    is_donated: boolean;
}

export interface ResponseMetadata {
    idType: string;
    value: string;
}

export interface Author {
    id?: Nullable<string>;
    name?: Nullable<string>;
}

export interface Publisher {
    id?: Nullable<string>;
    name?: Nullable<string>;
}

export interface Meta {
    ageRestriction?: Nullable<string>;
    coverType?: Nullable<string>;
    dimensions?: Nullable<string>;
    manufacturer?: Nullable<string>;
    mass?: Nullable<string>;
    numberOfPages?: Nullable<string>;
    price?: Nullable<string>;
    series?: Nullable<string>;
    sku?: Nullable<string>;
}

export interface IMaterialMeta {
    authors?: Nullable<Nullable<Author>[]>;
    cover?: Nullable<string>;
    description?: Nullable<string>;
    title?: Nullable<string>;
    yearPublishedAt?: Nullable<number>;
    meta?: Nullable<Meta>;
    publisher?: Nullable<Publisher>;
}

export interface IMetaOfMaterial {
    idType?: Nullable<string>;
    value?: Nullable<string>;
    material?: Nullable<IMaterialMeta>;
}

export interface Message {
    id: string;
    material_id?: Nullable<number>;
    person_id: number;
    created_at: DateTime;
    person: Person;
    material: Material;
    title: string;
    message: string;
    location_id: number;
}

export interface Notification {
    id: string;
    material_id?: Nullable<number>;
    person_id: number;
    created_at: DateTime;
    person: Person;
    material?: Nullable<Material>;
    message?: Nullable<string>;
}

export interface Person {
    id: string;
    smg_id: string;
    position: string;
    type: string;
    username: string;
    email: string;
    avatar: string;
    created_at: DateTime;
    permissions?: Nullable<Nullable<Permissions>[]>;
    statuses?: Nullable<Nullable<Status>[]>;
    notifications?: Nullable<Nullable<Notification>[]>;
    location?: Nullable<Location[]>;
    messages?: Nullable<Nullable<Message>[]>;
    states?: Nullable<Nullable<BlockedUsers>[]>;
}

export interface Status {
    id: string;
    material_id: number;
    person_id: number;
    status: string;
    created_at: DateTime;
    material: Material;
    person: Person;
}

export interface Error {
    message: string;
}

export type DateTime = any;
export type Time = any;
export type Timestamp = any;
export type TimeZone = any;
export type UtcOffset = any;
export type Duration = any;
export type ISO8601Duration = any;
export type LocalDate = any;
export type LocalTime = any;
export type LocalEndTime = any;
export type EmailAddress = any;
export type NegativeFloat = any;
export type NegativeInt = any;
export type NonEmptyString = any;
export type NonNegativeFloat = any;
export type NonNegativeInt = any;
export type NonPositiveFloat = any;
export type NonPositiveInt = any;
export type PhoneNumber = any;
export type PositiveFloat = any;
export type PositiveInt = any;
export type PostalCode = any;
export type UnsignedFloat = any;
export type UnsignedInt = any;
export type URL = any;
export type BigInt = any;
export type Long = any;
export type Byte = any;
export type UUID = any;
export type GUID = any;
export type Hexadecimal = any;
export type HexColorCode = any;
export type HSL = any;
export type HSLA = any;
export type IPv4 = any;
export type IPv6 = any;
export type ISBN = any;
export type JWT = any;
export type Latitude = any;
export type Longitude = any;
export type MAC = any;
export type Port = any;
export type RGB = any;
export type RGBA = any;
export type SafeInt = any;
export type USCurrency = any;
export type Currency = any;
export type JSON = any;
export type JSONObject = any;
export type IBAN = any;
export type ObjectID = any;
export type Void = any;
export type DID = any;
export type CountryCode = any;
export type Locale = any;
export type RoutingNumber = any;
export type AccountNumber = any;
export type BookUnionResult = Status | Error;
type Nullable<T> = T | null;
