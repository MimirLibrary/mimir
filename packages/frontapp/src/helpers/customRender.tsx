import '@testing-library/jest-dom';
import { render, RenderOptions } from '@testing-library/react';
import React, { FC, ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { ApolloProvider } from '@apollo/client';
import { client } from '@mimir/apollo-client';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string | null) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            return;
          }),
      },
    };
  },
}));

jest.unmock('react-redux');

const wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <BrowserRouter>
      <ApolloProvider client={client}>
        <Provider store={store}>{children}</Provider>
      </ApolloProvider>
    </BrowserRouter>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => {
  return render(ui, { wrapper, ...options });
};

export { customRender as render };
