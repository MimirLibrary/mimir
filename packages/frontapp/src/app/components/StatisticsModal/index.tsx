import styled from '@emotion/styled';
import { colors, dimensions } from '@mimir/ui-kit';
import { FC, Dispatch, SetStateAction, useEffect } from 'react';
import Modal from '../Modal';
import { t } from 'i18next';
import Button from '../Button';
import { useGetAllMaterialsForManagerQuery } from '@mimir/apollo-client';
import { useAppSelector } from '../../hooks/useTypedSelector';
import { locationIds } from '../../store/slices/userSlice';
import { toast } from 'react-toastify';
import { StatusTypes } from '@mimir/global-types';

interface StatisticsModal {
  isActive: boolean;
  setIsActive: Dispatch<SetStateAction<boolean>>;
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: ${dimensions.base_2};
`;

const Title = styled.h1`
  align-self: flex-start;
  font-weight: 700;
  font-size: ${dimensions.xl_2};
  line-height: ${dimensions.xl_3};
`;

const ButtonsWrapper = styled.div`
  position: absolute;
  bottom: 3rem;
  width: 90%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: ${dimensions.base};
`;

const PieChart = styled.figure<{
  free: number;
  claimed: number;
  overdue: number;
  pending: number;
}>`
  justify-self: center;
  background: radial-gradient(
      circle closest-side,
      ${colors.bg_secondary} 0,
      ${colors.bg_secondary} 55%,
      transparent 55%,
      transparent 100%
    ),
    conic-gradient(
      from 270deg,
      ${colors.main_green} 0,
      ${colors.main_green} ${(props) => props.free / 2}%,
      ${colors.accent_color} 0,
      ${colors.accent_color} ${(props) => props.free / 2 + props.claimed / 2}%,
      ${colors.problem_red} 0,
      ${colors.problem_red}
        ${(props) => props.free / 2 + props.claimed / 2 + props.overdue / 2}%,
      ${colors.gray_additional} 0,
      ${colors.gray_additional}
        ${(props) =>
          props.free / 2 +
          props.claimed / 2 +
          props.overdue / 2 +
          props.pending / 2}%,
      ${colors.bg_secondary} 0,
      ${colors.bg_secondary} 100%
    );
  width: 290px;
  height: 290px;
  border-radius: 50%;
  margin: 0;

  @media (max-width: ${dimensions.tablet_width}) {
    margin-bottom: 3rem;
  }
`;

const LegendWrapper = styled.div`
  position: absolute;
  top: 18rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: ${dimensions.xl_2};

  @media (max-width: ${dimensions.tablet_width}) {
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
    gap: ${dimensions.xs_2};
    margin-left: 3.5rem;
  } ;
`;

const LegendItem = styled.div`
  display: flex;
  align-items: center;
  gap: ${dimensions.xs_2};
`;

const LegendColor = styled.div`
  width: ${dimensions.base};
  height: ${dimensions.base};
  border-radius: 50%;
  background-color: ${({ color }) => color};
`;

const StatisticsModal: FC<StatisticsModal> = ({ isActive, setIsActive }) => {
  const locations = useAppSelector(locationIds);
  const { data, error } = useGetAllMaterialsForManagerQuery({
    variables: { locations },
    fetchPolicy: 'no-cache',
  });
  const bookItems = data?.getAllMaterials!.map((material) => {
    return { ...material, status: material?.statuses.slice(-1)[0] };
  });

  const NUMBER_OF_ALL_BOOKS =
    bookItems?.filter((item) => item.status?.status).length || 0;
  const NUMBER_OF_FREE_BOOKS =
    bookItems?.filter((item) => item?.status?.status === StatusTypes.FREE)
      .length || 0;
  const NUMBER_OF_CLAIMED_BOOKS =
    bookItems?.filter(
      (item) =>
        item?.status?.status === StatusTypes.BUSY ||
        item?.status?.status === StatusTypes.PROLONG
    ).length || 0;
  const NUMBER_OF_OVERDUE_BOOKS =
    bookItems?.filter((item) => item?.status?.status === StatusTypes.OVERDUE)
      .length || 0;
  const NUMBER_OF_PENDING_BOOKS =
    bookItems?.filter((item) => item?.status?.status === StatusTypes.PENDING)
      .length || 0;

  const STATISTICS = {
    free: Math.round((NUMBER_OF_FREE_BOOKS / NUMBER_OF_ALL_BOOKS) * 100),
    claimed: Math.round((NUMBER_OF_CLAIMED_BOOKS / NUMBER_OF_ALL_BOOKS) * 100),
    overdue: Math.round((NUMBER_OF_OVERDUE_BOOKS / NUMBER_OF_ALL_BOOKS) * 100),
    pending: Math.round((NUMBER_OF_PENDING_BOOKS / NUMBER_OF_ALL_BOOKS) * 100),
  };

  const closeModal = () => {
    setIsActive(false);
  };

  useEffect(() => {
    if (error) {
      toast.error(error.message);
    }
  }, [error]);

  return (
    <Modal active={isActive} setActive={setIsActive}>
      <Wrapper>
        <Title>{t('Statistics.Title')}</Title>
        <PieChart
          free={STATISTICS.free}
          claimed={STATISTICS.claimed}
          overdue={STATISTICS.overdue}
          pending={STATISTICS.pending}
          data-testid="statistics-chart"
        ></PieChart>
        <LegendWrapper data-testid="statistics-legend">
          <LegendItem data-testid="statistics-free">
            <LegendColor color={colors.main_green}></LegendColor>
            <span>
              {t('Statuses.Free')} - {NUMBER_OF_FREE_BOOKS}
            </span>
          </LegendItem>
          <LegendItem data-testid="statistics-claimed">
            <LegendColor color={colors.accent_color}></LegendColor>
            <span>
              {t('Statuses.Claimed')} - {NUMBER_OF_CLAIMED_BOOKS}
            </span>
          </LegendItem>
          <LegendItem data-testid="statistics-overdue">
            <LegendColor color={colors.problem_red}></LegendColor>
            <span>
              {t('Statuses.Overdue')} - {NUMBER_OF_OVERDUE_BOOKS}
            </span>
          </LegendItem>
          <LegendItem data-testid="statistics-pending">
            <LegendColor color={colors.gray_additional}></LegendColor>
            <span>
              {t('Statuses.Pending')} - {NUMBER_OF_PENDING_BOOKS}
            </span>
          </LegendItem>
        </LegendWrapper>
        <ButtonsWrapper>
          <Button transparent value="Cancel" onClick={closeModal} />
          <Button value="Ok" onClick={closeModal} />
        </ButtonsWrapper>
      </Wrapper>
    </Modal>
  );
};

export default StatisticsModal;
