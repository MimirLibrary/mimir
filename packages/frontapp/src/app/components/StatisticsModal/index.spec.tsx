import { MockedProvider } from '@apollo/client/testing';
import { GetAllMaterialsForManagerDocument } from '@mimir/apollo-client';
import { StatusTypes } from '@mimir/global-types';
import { render } from 'packages/frontapp/src/helpers/customRender';
import StatisticsModal from './index';
import { screen } from '@testing-library/react';
import { nextTick } from '../../../helpers/tests/nextTick';

const materialsMock = {
  request: {
    query: GetAllMaterialsForManagerDocument,
    variables: { input: { locations: [] } },
  },
  result: {
    data: {
      getAllMaterials: [
        {
          category: 'History',
          id: '1',
          picture: '',
          currentStatus: { id: '7', status: StatusTypes.FREE },
          title: 'Forrest Gump',
        },
        {
          category: 'Psychology',
          id: '2',
          picture: '',
          currentStatus: { id: '8', status: StatusTypes.FREE },
          title: 'New Patient',
        },
        {
          category: 'Psychology',
          id: '3',
          picture: '',
          currentStatus: { id: '10', status: StatusTypes.BUSY },
          title: 'Charity',
        },
        {
          category: 'History',
          id: '4',
          picture: '',
          currentStatus: { id: '9', status: StatusTypes.OVERDUE },
          title: 'Public History',
        },
      ],
    },
  },
};

jest.mock('i18next', () => ({
  t: (i: any) => i,
}));

describe('Statistics Modal Component', () => {
  beforeEach(() => {
    const handleCloseModal = jest.fn();
    render(
      <MockedProvider mocks={[materialsMock]} addTypename={false}>
        <StatisticsModal isActive={true} setIsActive={handleCloseModal} />
      </MockedProvider>
    );
  });

  it('should render correct statistics', async () => {
    await nextTick();

    expect(screen.getByTestId('statistics-chart')).toBeInTheDocument();
    expect(screen.getByTestId('statistics-free')).toHaveTextContent('2');
    expect(screen.getByTestId('statistics-claimed')).toHaveTextContent('1');
    expect(screen.getByTestId('statistics-overdue')).toHaveTextContent('1');
    expect(screen.getByTestId('statistics-pending')).toHaveTextContent('0');
  });

  it('should match snapshot', async () => {
    await nextTick();

    expect(screen.getByTestId('statistics-chart')).toMatchSnapshot();
    expect(screen.getByTestId('statistics-legend')).toMatchSnapshot();
  });
});
