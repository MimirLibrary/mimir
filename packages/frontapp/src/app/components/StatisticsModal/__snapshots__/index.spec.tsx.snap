// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics Modal Component should match snapshot 1`] = `
.emotion-0 {
  justify-self: center;
  background: radial-gradient(
        circle closest-side,
        #FFFFFF 0,
        #FFFFFF 55%,
        transparent 55%,
        transparent 100%
      ),conic-gradient(
        from 270deg,
        #2AA816 0,
        #2AA816 25%,
        #1A1ED6 0,
        #1A1ED6 37.5%,
        #FF372A 0,
        #FF372A
          50%,
        #E0E0E0 0,
        #E0E0E0
          50%,
        #FFFFFF 0,
        #FFFFFF 100%
      );
  width: 290px;
  height: 290px;
  border-radius: 50%;
  margin-bottom: -10rem;
  margin-top: -10px;
}

<figure
  class="emotion-0"
  data-testid="statistics-chart"
/>
`;

exports[`Statistics Modal Component should match snapshot 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  top: 18rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  gap: 1.5rem;
  margin-bottom: -10px;
}

@media (max-width: 48rem) {
  .emotion-0 {
    display: grid;
    grid-template-columns: 1fr 1fr;
    row-gap: 1rem;
    margin-left: 2rem;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.5rem;
}

.emotion-2 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #2AA816;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: 2px;
}

.emotion-4 {
  font: #333333;
  font-size: 0.875rem;
  line-height: 16px;
}

.emotion-5 {
  font: #333333;
  font-size: 0.75rem;
  line-height: 16px;
}

.emotion-7 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #1A1ED6;
}

.emotion-12 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FF372A;
}

.emotion-17 {
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #E0E0E0;
}

<div
  class="emotion-0"
  data-testid="statistics-legend"
>
  <div
    class="emotion-1"
    data-testid="statistics-free"
  >
    <div
      class="emotion-2"
      color="#2AA816"
    />
    <div
      class="emotion-3"
    >
      <div
        class="emotion-4"
      >
        Statuses.Free
      </div>
      <div
        class="emotion-5"
      >
        2
      </div>
    </div>
  </div>
  <div
    class="emotion-1"
    data-testid="statistics-claimed"
  >
    <div
      class="emotion-7"
      color="#1A1ED6"
    />
    <div
      class="emotion-3"
    >
      <div
        class="emotion-4"
      >
        Statuses.Claimed
      </div>
      <div
        class="emotion-5"
      >
        1
      </div>
    </div>
  </div>
  <div
    class="emotion-1"
    data-testid="statistics-overdue"
  >
    <div
      class="emotion-12"
      color="#FF372A"
    />
    <div
      class="emotion-3"
    >
      <div
        class="emotion-4"
      >
        Statuses.Overdue
      </div>
      <div
        class="emotion-5"
      >
        1
      </div>
    </div>
  </div>
  <div
    class="emotion-1"
    data-testid="statistics-pending"
  >
    <div
      class="emotion-17"
      color="#E0E0E0"
    />
    <div
      class="emotion-3"
    >
      <div
        class="emotion-4"
      >
        Statuses.Pending
      </div>
      <div
        class="emotion-5"
      >
        0
      </div>
    </div>
  </div>
</div>
`;
